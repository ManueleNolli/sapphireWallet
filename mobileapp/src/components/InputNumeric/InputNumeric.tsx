import { Input, StyleService, Text, useStyleSheet } from '@ui-kitten/components'
import React, { useEffect } from 'react'
import { TouchableOpacity } from 'react-native'

type InputAddressProps = {
  value: string
  setValue: (value: string) => void
  isValid: boolean
  setIsValid: (isValid: boolean) => void
  maxValue?: number
  style?: any
}

export default function InputNumeric({
  value,
  setValue,
  isValid,
  setIsValid,
  maxValue,
  style,
}: InputAddressProps) {
  const styles = useStyleSheet(themedStyles)

  useEffect(() => {
    const valueNumber = Number.parseFloat(value)
    setIsValid(valueNumber >= 0 && (!maxValue || valueNumber <= maxValue))
  }, [value, maxValue])

  const renderMaxValueButton = () => {
    if (maxValue) {
      return (
        <TouchableOpacity
          testID="max-value-button"
          onPress={() => setValue(maxValue.toString())}
        >
          <Text style={styles.textMaxValue} category={'label'}>
            Max Value
          </Text>
        </TouchableOpacity>
      )
    }
    return null
  }

  const status = () => {
    if (isValid) {
      return 'success'
    } else if (value.length > 0) {
      return 'danger'
    }
    return 'basic'
  }

  return (
    <Input
      testID={'input-numeric'}
      value={value}
      label="Amount"
      placeholder="0"
      style={style}
      keyboardType="numeric"
      status={status()}
      onChangeText={(nextValue: string) => setValue(nextValue)}
      accessoryRight={<>{renderMaxValueButton()}</>}
    />
  )
}

const themedStyles = StyleService.create({
  textMaxValue: {
    color: 'color-primary-500',
  },
})
