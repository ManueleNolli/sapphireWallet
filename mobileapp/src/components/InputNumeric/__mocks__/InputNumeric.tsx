const InputNumericMock = () => (
  <div data-testid={'InputNumeric'}>InputNumeric</div>
)

export default InputNumericMock
