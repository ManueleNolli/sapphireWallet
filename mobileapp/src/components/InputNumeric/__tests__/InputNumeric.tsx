import React from 'react'
import { act, fireEvent, waitFor } from '@testing-library/react-native'
import InputNumeric from '../InputNumeric'
import renderWithTheme from '../../../TestHelper'

describe('InputNumeric Component', () => {
  it('renders correctly', () => {
    const tree = renderWithTheme(
      <InputNumeric
        value=""
        setValue={() => {}}
        isValid={false}
        setIsValid={() => {}}
      />
    )

    expect(tree.getAllByText('Amount')).toBeTruthy()
    expect(tree.getByPlaceholderText('0')).toBeTruthy()
    expect(tree).toMatchSnapshot()
  })

  it('updates value and calls setIsValid correctly', async () => {
    const setIsValidMock = jest.fn()

    let tree: any

    await waitFor(async () => {
      tree = renderWithTheme(
        <InputNumeric
          value=""
          setValue={() => {}}
          isValid={false}
          setIsValid={setIsValidMock}
        />
      )
    })

    await act(async () => {
      const button = tree!.getAllByText('Amount')
      fireEvent.changeText(button[0], '42')
    })

    expect(setIsValidMock).toHaveBeenCalled()
  })

  it('displays correct status', () => {
    const { getByTestId } = renderWithTheme(
      <InputNumeric
        value=""
        setValue={() => {}}
        isValid={false}
        setIsValid={() => {}}
      />
    )

    expect(getByTestId('@input-numeric/input').props.status).toBe('basic')
  })

  it('displays success status when isValid is true', () => {
    const { getByTestId } = renderWithTheme(
      <InputNumeric
        value=""
        setValue={() => {}}
        isValid={true}
        setIsValid={() => {}}
      />
    )

    expect(getByTestId('@input-numeric/input').props.status).toBe('success')
  })

  it('displays danger status when isValid is false and value length is greater than 0', () => {
    const { getByTestId } = renderWithTheme(
      <InputNumeric
        value="invalidValue"
        setValue={() => {}}
        isValid={false}
        setIsValid={() => {}}
      />
    )

    expect(getByTestId('@input-numeric/input').props.status).toBe('danger')
  })

  it('calls renderMaxValueButton onPress correctly', () => {
    const setValueMock = jest.fn()
    const { getByTestId } = renderWithTheme(
      <InputNumeric
        value=""
        setValue={setValueMock}
        isValid={false}
        setIsValid={() => {}}
        maxValue={100}
      />
    )

    fireEvent.press(getByTestId('max-value-button'))

    expect(setValueMock).toHaveBeenCalledWith('100')
  })
})
