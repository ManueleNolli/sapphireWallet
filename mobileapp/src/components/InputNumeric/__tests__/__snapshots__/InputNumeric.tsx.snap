// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputNumeric Component renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="@input-numeric/container"
>
  <Text
    appearance="default"
    category="p1"
    status="basic"
    style={
      [
        {
          "color": "#222B45",
          "fontFamily": "System",
          "fontSize": 15,
          "fontWeight": "400",
        },
        [
          {
            "color": "#8F9BB3",
            "fontFamily": "System",
            "fontSize": 12,
            "fontWeight": "800",
            "marginBottom": 4,
          },
          {
            "textAlign": "left",
          },
        ],
      ]
    }
  >
    Amount
  </Text>
  <View
    style={
      [
        {
          "backgroundColor": "#F7F9FC",
          "borderColor": "#E4E9F2",
          "borderRadius": 4,
          "borderWidth": 1,
          "minHeight": 40,
          "paddingHorizontal": 8,
          "paddingVertical": 7,
        },
        {
          "alignItems": "center",
          "flexDirection": "row",
          "width": "100%",
        },
      ]
    }
  >
    <TextInput
      appearance="default"
      editable={true}
      keyboardType="numeric"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      placeholder="0"
      placeholderTextColor="#8F9BB3"
      size="medium"
      status="basic"
      style={
        [
          {
            "color": "#222B45",
            "fontFamily": "System",
            "fontSize": 15,
            "fontWeight": "normal",
            "marginHorizontal": 8,
          },
          {
            "flexBasis": "auto",
            "flexGrow": 1,
            "flexShrink": 1,
          },
          null,
          undefined,
        ]
      }
      testID="@input-numeric/input"
      value=""
    />
  </View>
</View>
`;
